src/System.Private.CoreLib/shared/System/MathF.cs

```diff
@@ -36,6 +36,76 @@ public static float Abs(float x)
             return Math.Abs(x);
         }
 
+        public static float BitDecrement(float x)
+        {
+            var bits = BitConverter.SingleToInt32Bits(x);
+
+            if ((bits & 0x7F800000) >= 0x7F800000)
+            {
+                // NaN returns NaN
+                // -Infinity returns -Infinity
+                // +Infinity returns float.MaxValue
+                return (bits == 0x7F800000) ? float.MaxValue : x;
+            }
+
+            if (bits == 0x00000000)
+            {
+                // +0.0 returns -float.Epsilon
+                return -float.Epsilon;
+            }
+
+            // Negative values need to be incremented
+            // Positive values need to be decremented
+
+            bits += ((bits < 0) ? +1 : -1);
+            return BitConverter.Int32BitsToSingle(bits);
+        }
+
+        public static float BitIncrement(float x)
+        {
+            var bits = BitConverter.SingleToInt32Bits(x);
+
+            if ((bits & 0x7F800000) >= 0x7F800000)
+            {
+                // NaN returns NaN
+                // -Infinity returns float.MinValue
+                // +Infinity returns +Infinity
+                return (bits == unchecked((int)(0xFF800000))) ? float.MinValue : x;
+            }
+
+            if (bits == unchecked((int)(0x80000000)))
+            {
+                // -0.0 returns float.Epsilon
+                return float.Epsilon;
+            }
+
+            // Negative values need to be decremented
+            // Positive values need to be incremented
+
+            bits += ((bits < 0) ? -1 : +1);
+            return BitConverter.Int32BitsToSingle(bits);
+        }
+
+        public static unsafe float CopySign(float x, float y)
```

Comment:
Was looking at the codegen for this method:
```asm
; Assembly listing for method MathF:CopySign(float,float):float
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T03] (  4,  3.50)   float  ->  mm0        
;  V01 arg1         [V01,T04] (  3,  3   )   float  ->  mm1        
;  V02 loc0         [V02,T00] (  3,  2.50)     int  ->  rax        
;  V03 loc1         [V03,T01] (  2,  2   )     int  ->  rdx        
;# V04 OutArgs      [V04    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]   "OutgoingArgSpace"
;  V05 tmp1         [V05,T05] (  2,  4   )   float  ->  [rsp+0x14]   do-not-enreg[F] ld-addr-op "Inlining Arg"
;  V06 tmp2         [V06,T06] (  2,  4   )   float  ->  [rsp+0x10]   do-not-enreg[F] ld-addr-op "Inlining Arg"
;  V07 tmp3         [V07,T02] (  2,  2   )     int  ->  [rsp+0x0C]   do-not-enreg[F] ld-addr-op "Inlining Arg"
;
; Lcl frame size = 24

G_M7953_IG01:
       sub      rsp, 24
       vzeroupper 

G_M7953_IG02:
       vmovss   dword ptr [rsp+14H], xmm0
       mov      eax, dword ptr [rsp+14H]
       vmovss   dword ptr [rsp+10H], xmm1
       mov      edx, dword ptr [rsp+10H]
       xor      edx, eax
       test     edx, edx
       jge      SHORT G_M7953_IG04
       xor      eax, 0xD1FFAB1E
       mov      dword ptr [rsp+0CH], eax
       vmovss   xmm0, dword ptr [rsp+0CH]

G_M7953_IG03:
       add      rsp, 24
       ret      

G_M7953_IG04:
       add      rsp, 24
       ret      

; Total bytes of code 61, prolog size 7 for method MathF:CopySign(float,float):float
; ============================================================
```

Reply:
The following, which is produced from `BitConverter.SingleToInt32Bits`
```asm
vmovss   dword ptr [rsp+14H], xmm0
mov      eax, dword ptr [rsp+14H]
```

Could be:
```asm
movd eax, xmm0
```

Reply:
The following, which is produced by `if ((xbits ^ ybits) < 0)`:
```asm
xor      edx, eax
test     edx, edx
jge      SHORT G_M7953_IG04
```

Could be:
```asm
xor edx, eax
js SHORT G_M7953_IG04
```

Reply:
`G_M7953_IG03:` and `G_M7953_IG04:` are the same, and could be folded together.

Reply:
The following, which is produced from BitConverter.Int32BitsToSingle
```asm
mov      dword ptr [rsp+0CH], eax
vmovss   xmm0, dword ptr [rsp+0CH]
```

Could be:
```asm
vmovd xmm0, eax
```

Reply:
Based on the April 2018 edition of the `Intel 64 and IA-32 Architectures Optimization Reference Manual`, it doesn't look like we need `vzeroupper`, as it specifies there is no transition penalty when using the AVX encoded 128-bit instructions, only when you are executing 256-bit instructions:
![image](https://user-images.githubusercontent.com/10487869/47966290-8f64b400-e005-11e8-88dc-899a2c6eb482.png)

Perhaps @fiigii could clarify here?

Reply:
This means we should be generating something closer to the following, if possible:
```asm
G_M7953_IG01:
       vmovd    eax, xmm0
       vmovd    edx, xmm1
       xor      edx, eax
       jns      SHORT G_M7953_IG02
       xor      eax, 0xD1FFAB1E
       vmovd    xmm0, eax

G_M7953_IG02:
       ret
```

Reply:
> as it specifies there is no transition penalty when using the AVX encoded 128-bit instructions,

Right, there is no transition penalty in VEX encoding, but we insert `vzeroupper` at the prolog to avoid SSE-to-AVX transition penalty from P/Invoke or corssgen-ed code calling JITed code.

Reply:
> at the prolog to avoid SSE-to-AVX transition penalty from P/Invoke or corssgen-ed code calling JITed code.

The architecture optimization manual seems to imply that there is no transition penalty when going from legacy 128-bit instructions to VEX-encoded 128-bit instructions (or vice-versa). Only when going from legacy 128-bit instructions to VEX-encoded 256-bit instructions (and that you avoid the penalty by using `VZEROUPPER`) and when going from VEX-encoded 256-bit instructions to legacy 128-bit instructions (where you cannot avoid the penalty). -- See the last line in the table, for `Mixed Code Handling` (and some surrounding text in the optimization manual)

Reply:
Ah, the optimization manual is confusing here. Yes, there is no transition penalty when going from legacy 128-bit instructions to VEX-encoded 128-bit instructions on **Skylake (and newer) micro-architectures**.
![image](https://user-images.githubusercontent.com/1263030/48019732-f73ffb00-e0e8-11e8-9159-910c78991927.png)

But the penalty still can happen on older architectures
![image](https://user-images.githubusercontent.com/1263030/48019799-248ca900-e0e9-11e8-986f-6e0deefec298.png)

Reply:
Ah, I see. There is a penalty on Broadwell and prior when going from `Preserved Non-INIT Upper State` to `Dirty Upper State` or `Clean Upper State`. However, it looks like `Penalty D` and `Penalty B` are approx the same (whether it is implicit to dirty, or explicit to `clean` via `vzeroupper`). The difference is whether there is an additional penalty when transitioning back into the legacy instructions (which will impact crossgen code)...

![image](https://user-images.githubusercontent.com/10487869/48020032-c3190a00-e0e9-11e8-9e21-c6856723b10d.png)
![image](https://user-images.githubusercontent.com/10487869/48020041-c7452780-e0e9-11e8-9f23-2eee411a7558.png)
![image](https://user-images.githubusercontent.com/10487869/48020094-e17f0580-e0e9-11e8-8b8d-cfd21efe8040.png)